File: core/src/main/java/ch/dissem/bitmessage/entity/valueobject/PrivateKey.java
/*
 * Copyright 2015 Christian Basler
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ch.dissem.bitmessage.entity.valueobject;

import ch.dissem.bitmessage.entity.BitmessageAddress;
import ch.dissem.bitmessage.entity.Streamable;
import ch.dissem.bitmessage.entity.payload.Pubkey;
import ch.dissem.bitmessage.exception.ApplicationException;
import ch.dissem.bitmessage.factory.Factory;
import ch.dissem.bitmessage.utils.Bytes;
import ch.dissem.bitmessage.utils.Decode;
import ch.dissem.bitmessage.utils.Encode;

import java.io.*;
import java.util.ArrayList;
import java.util.List;

import static ch.dissem.bitmessage.utils.Singleton.security;

/**
 * Represents a private key. Additional information (stream, version, features, ...) is stored in the accompanying
 * {@link Pubkey} object.
 */
public class PrivateKey implements Streamable {
    private static final long serialVersionUID = 8562555470709110558L;

    public static final int PRIVATE_KEY_SIZE = 32;

    private final byte[] privateSigningKey;
    private final byte[] privateEncryptionKey;

    private final Pubkey pubkey;

    public PrivateKey(boolean shorter, long stream, long nonceTrialsPerByte, long extraBytes, Pubkey.Feature... features) {
        byte[] privSK;
        byte[] pubSK;
        byte[] privEK;
        byte[] pubEK;
        byte[] ripe;
        do {
            privSK = security().randomBytes(PRIVATE_KEY_SIZE);
            privEK = security().randomBytes(PRIVATE_KEY_SIZE);
            pubSK = security().createPublicKey(privSK);
            pubEK = security().createPublicKey(privEK);
            ripe = Pubkey.getRipe(pubSK, pubEK);
        } while (ripe[0] != 0 || (shorter && ripe[1] != 0));
        this.privateSigningKey = privSK;
        this.privateEncryptionKey = privEK;
        this.pubkey = security().createPubkey(Pubkey.LATEST_VERSION, stream, privateSigningKey, privateEncryptionKey,
                nonceTrialsPerByte, extraBytes, features);
    }

    public PrivateKey(byte[] privateSigningKey, byte[] privateEncryptionKey, Pubkey pubkey) {
        this.privateSigningKey = privateSigningKey;
        this.privateEncryptionKey = privateEncryptionKey;
        this.pubkey = pubkey;
    }

    public PrivateKey(BitmessageAddress address, String passphrase) {
        this(address.getVersion(), address.getStream(), passphrase);
    }

    public PrivateKey(long version, long stream, String passphrase) {
        this(new Builder(version, stream, false).seed(passphrase).generate());
    }

    private PrivateKey(Builder builder) {
        this.privateSigningKey = builder.privSK;
        this.privateEncryptionKey = builder.privEK;
        this.pubkey = Factory.createPubkey(builder.version, builder.stream, builder.pubSK, builder.pubEK, 0, 0);
    }

    private static class Builder {
        final long version;
        final long stream;
        final boolean shorter;

        byte[] seed;
        long nextNonce;

        byte[] privSK, privEK;
        byte[] pubSK, pubEK;

        private Builder(long version, long stream, boolean shorter) {
            this.version = version;
            this.stream = stream;
            this.shorter = shorter;
        }

        Builder seed(String passphrase) {
            try {
                seed = passphrase.getBytes("UTF-8");
            } catch (UnsupportedEncodingException e) {
                throw new ApplicationException(e);
            }
            return this;
        }

        Builder generate() {
            try {
                long signingKeyNonce = nextNonce;
                long encryptionKeyNonce = nextNonce + 1;
                byte[] ripe;
                do {
                    privEK = Bytes.truncate(security().sha512(seed, Encode.varInt(encryptionKeyNonce)), 32);
                    privSK = Bytes.truncate(security().sha512(seed, Encode.varInt(signingKeyNonce)), 32);
                    pubSK = security().createPublicKey(privSK);
                    pubEK = security().createPublicKey(privEK);
                    ripe = security().ripemd160(security().sha512(pubSK, pubEK));

                    signingKeyNonce += 2;
                    encryptionKeyNonce += 2;
                } while (ripe[0] != 0 || (shorter && ripe[1] != 0));
                nextNonce = signingKeyNonce;
            } catch (IOException e) {
                throw new ApplicationException(e);
            }
            return this;
        }
    }

    public static List<PrivateKey> deterministic(String passphrase, int numberOfAddresses, long version, long stream, boolean shorter) {
        List<PrivateKey> result = new ArrayList<>(numberOfAddresses);
        Builder builder = new Builder(version, stream, shorter).seed(passphrase);
        for (int i = 0; i < numberOfAddresses; i++) {
            builder.generate();
            result.add(new PrivateKey(builder));
        }
        return result;
    }

    public static PrivateKey read(InputStream is) throws IOException {
        int version = (int) Decode.varInt(is);
        long stream = Decode.varInt(is);
        int len = (int) Decode.varInt(is);
        Pubkey pubkey = Factory.readPubkey(version, stream, is, len, false);
        len = (int) Decode.varInt(is);
        byte[] signingKey = Decode.bytes(is, len);
        len = (int) Decode.varInt(is);
        byte[] encryptionKey = Decode.bytes(is, len);
        return new PrivateKey(signingKey, encryptionKey, pubkey);
    }

    public byte[] getPrivateSigningKey() {
        return privateSigningKey;
    }

    public byte[] getPrivateEncryptionKey() {
        return privateEncryptionKey;
    }

    public Pubkey getPubkey() {
        return pubkey;
    }

    @Override
    public void write(OutputStream out) throws IOException {
        Encode.varInt(pubkey.getVersion(), out);
        Encode.varInt(pubkey.getStream(), out);
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        pubkey.writeUnencrypted(baos);
        Encode.varInt(baos.size(), out);
        out.write(baos.toByteArray());
        Encode.varInt(privateSigningKey.length, out);
        out.write(privateSigningKey);
        Encode.varInt(privateEncryptionKey.length, out);
        out.write(privateEncryptionKey);
    }
}

File: core/src/main/kotlin/ch/dissem/bitmessage/entity/valueobject/PrivateKey.kt
/*
 * Copyright 2017 Christian Basler
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ch.dissem.bitmessage.entity.valueobject

import ch.dissem.bitmessage.InternalContext
import ch.dissem.bitmessage.entity.BitmessageAddress
import ch.dissem.bitmessage.entity.Streamable
import ch.dissem.bitmessage.entity.StreamableWriter
import ch.dissem.bitmessage.entity.payload.Pubkey
import ch.dissem.bitmessage.exception.ApplicationException
import ch.dissem.bitmessage.factory.Factory
import ch.dissem.bitmessage.utils.Bytes
import ch.dissem.bitmessage.utils.Decode
import ch.dissem.bitmessage.utils.Encode
import ch.dissem.bitmessage.utils.Singleton.cryptography
import java.io.*
import java.nio.ByteBuffer
import java.util.*

/**
 * Represents a private key. Additional information (stream, version, features, ...) is stored in the accompanying
 * [Pubkey] object.
 */
data class PrivateKey(
    val privateSigningKey: ByteArray,
    val privateEncryptionKey: ByteArray,

    val pubkey: Pubkey
) : Streamable {

    constructor(
        shorter: Boolean,
        stream: Long,
        nonceTrialsPerByte: Long, extraBytes: Long,
        vararg features: Pubkey.Feature
    ) : this(
        Builder(version = Pubkey.LATEST_VERSION, stream = stream, shorter = shorter)
            .random()
            .nonceTrialsPerByte(nonceTrialsPerByte)
            .extraBytes(extraBytes)
            .features(features)
            .generate())

    constructor(address: BitmessageAddress, passphrase: String) : this(address.version, address.stream, passphrase)

    constructor(version: Long, stream: Long, passphrase: String) : this(
        Builder(version, stream, false).seed(passphrase).generate()
    )

    private constructor(builder: Builder) : this(
        builder.privSK!!, builder.privEK!!,
        Factory.createPubkey(builder.version, builder.stream, builder.pubSK!!, builder.pubEK!!,
            builder.nonceTrialsPerByte, builder.extraBytes, *builder.features)
    )

    override fun equals(other: Any?) = other is PrivateKey
        && Arrays.equals(privateEncryptionKey, other.privateEncryptionKey)
        && Arrays.equals(privateSigningKey, other.privateSigningKey)
        && pubkey == other.pubkey

    override fun hashCode() = pubkey.hashCode()

    override fun writer(): StreamableWriter = Writer(this)

    private class Writer(
        private val item: PrivateKey
    ) : StreamableWriter {

        override fun write(out: OutputStream) {
            Encode.varInt(item.pubkey.version, out)
            Encode.varInt(item.pubkey.stream, out)
            val baos = ByteArrayOutputStream()
            item.pubkey.writer().writeUnencrypted(baos)
            Encode.varInt(baos.size(), out)
            out.write(baos.toByteArray())
            Encode.varBytes(item.privateSigningKey, out)
            Encode.varBytes(item.privateEncryptionKey, out)
        }

        override fun write(buffer: ByteBuffer) {
            Encode.varInt(item.pubkey.version, buffer)
            Encode.varInt(item.pubkey.stream, buffer)
            try {
                val baos = ByteArrayOutputStream()
                item.pubkey.writer().writeUnencrypted(baos)
                Encode.varBytes(baos.toByteArray(), buffer)
            } catch (e: IOException) {
                throw ApplicationException(e)
            }

            Encode.varBytes(item.privateSigningKey, buffer)
            Encode.varBytes(item.privateEncryptionKey, buffer)
        }

    }

    private class Builder internal constructor(
        internal val version: Long,
        internal val stream: Long,
        internal val shorter: Boolean
    ) {

        internal var seed: ByteArray? = null
        internal var nextNonce: Long = 0

        internal var privSK: ByteArray? = null
        internal var privEK: ByteArray? = null
        internal var pubSK: ByteArray? = null
        internal var pubEK: ByteArray? = null

        internal var nonceTrialsPerByte = InternalContext.NETWORK_NONCE_TRIALS_PER_BYTE
        internal var extraBytes = InternalContext.NETWORK_EXTRA_BYTES
        internal var features: Array<out Pubkey.Feature> = emptyArray()

        internal fun random(): Builder {
            seed = cryptography().randomBytes(1024)
            return this
        }

        fun nonceTrialsPerByte(nonceTrialsPerByte: Long): Builder {
            this.nonceTrialsPerByte = nonceTrialsPerByte
            return this
        }

        fun extraBytes(extraBytes: Long): Builder {
            this.extraBytes = extraBytes
            return this
        }

        fun features(features: Array<out Pubkey.Feature>): Builder {
            this.features = features
            return this
        }

        internal fun seed(passphrase: String): Builder {
            try {
                seed = passphrase.toByteArray(charset("UTF-8"))
            } catch (e: UnsupportedEncodingException) {
                throw ApplicationException(e)
            }

            return this
        }

        internal fun generate(): Builder {
            var signingKeyNonce = nextNonce
            var encryptionKeyNonce = nextNonce + 1
            var ripe: ByteArray
            do {
                privEK = Bytes.truncate(cryptography().sha512(seed!!, Encode.varInt(encryptionKeyNonce)), 32)
                privSK = Bytes.truncate(cryptography().sha512(seed!!, Encode.varInt(signingKeyNonce)), 32)
                pubSK = cryptography().createPublicKey(privSK!!)
                pubEK = cryptography().createPublicKey(privEK!!)
                ripe = cryptography().ripemd160(cryptography().sha512(pubSK!!, pubEK!!))

                signingKeyNonce += 2
                encryptionKeyNonce += 2
            } while (ripe[0].toInt() != 0 || shorter && ripe[1].toInt() != 0)
            nextNonce = signingKeyNonce
            return this
        }
    }

    companion object {
        @JvmField
        val PRIVATE_KEY_SIZE = 32

        @JvmStatic
        fun deterministic(passphrase: String, numberOfAddresses: Int, version: Long, stream: Long, shorter: Boolean): List<PrivateKey> {
            val result = ArrayList<PrivateKey>(numberOfAddresses)
            val builder = Builder(version, stream, shorter).seed(passphrase)
            for (i in 0..numberOfAddresses - 1) {
                builder.generate()
                result.add(PrivateKey(builder))
            }
            return result
        }

        @JvmStatic
        fun read(input: InputStream): PrivateKey {
            val version = Decode.varInt(input).toInt()
            val stream = Decode.varInt(input)
            val len = Decode.varInt(input).toInt()
            val pubkey = Factory.readPubkey(version.toLong(), stream, input, len, false) ?: throw ApplicationException("Unknown pubkey version encountered")
            val signingKey = Decode.varBytes(input)
            val encryptionKey = Decode.varBytes(input)
            return PrivateKey(signingKey, encryptionKey, pubkey)
        }
    }
}
